File: src/filter.c
/*  Copyright (C) 2012, 2013 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "filter.h"
#include "stringarray.h"

#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <regex.h>


typedef struct regex_list_t
{
    size_t    length;
    regex_t * regexs;
    size_t    total_count;
    size_t    match_count;
} regex_list_t;

static void compile(config_setting_t const * array, regex_list_t * prepared);
static int  match(regex_list_t * prepared, char const * input);
static int  is_empty(regex_list_t const * prepared);
static void release(regex_list_t * prepared);

static config_setting_t const * get_setting(config_setting_t const *, char const *);


struct bear_output_filter_t
{
    regex_list_t compilers;
    regex_list_t source_files;
    regex_list_t cancel_parameters;
};


static char const * fix_path(char const * file, char const * cwd);

bear_output_filter_t * bear_filter_read_from_file(char const * file)
{
    config_t config;
    config_init(&config);
    if (config_read_file(&config, file) == CONFIG_FALSE)
    {
        fprintf(stderr, "bear: failed to configure: '%s' in file %s at line %d\n",
                config_error_text(&config),
                config_error_file(&config),
                config_error_line(&config));
        exit(EXIT_FAILURE);
    }
    bear_output_filter_t * const result = bear_filter_create(&config);
    config_destroy(&config);

    return result;
}

bear_output_filter_t * bear_filter_create(config_t const * config)
{
    bear_output_filter_t * filter = malloc(sizeof(bear_output_filter_t));
    if (0 == filter)
    {
        perror("bear: malloc");
        exit(EXIT_FAILURE);
    }

    config_setting_t * const group = config_lookup(config, "filter");
    if (0 == group)
    {
        fprintf(stderr, "bear: found no filter group in config file.\n");
        exit(EXIT_FAILURE);
    }

    compile(get_setting(group, "compilers"), &filter->compilers);
    compile(get_setting(group, "source_files"), &filter->source_files);
    compile(get_setting(group, "cancel_parameters"), &filter->cancel_parameters);

    if (is_empty(&filter->compilers) || is_empty(&filter->source_files))
    {
        fprintf(stderr, "bear: empty compilers or source files in config file will produce empty output.\n");
        exit(EXIT_FAILURE);
    }

    return filter;
}

void bear_filter_report(bear_output_filter_t * filter)
{
    if (0 == filter)
        fprintf(stderr, "bear: filtering were not enabled.\n");
    else
    {
        fprintf(stderr, "bear: filtering statistic:\n");
        fprintf(stderr, "  total number of child processes : %zu\n", filter->compilers.total_count);
        fprintf(stderr, "  match as compiler               : %zu\n", filter->compilers.match_count);
        fprintf(stderr, "  match as source file            : %zu\n", filter->source_files.match_count);
        fprintf(stderr, "  match on cancel parameter       : %zu\n", filter->cancel_parameters.match_count);
    }
}

void bear_filter_delete(bear_output_filter_t * filter)
{
    if (0 == filter)
        return;

    release(&filter->compilers);
    release(&filter->source_files);
    release(&filter->cancel_parameters);

    free((void *)filter);
}

char const * bear_filter_source_file(bear_output_filter_t * filter, bear_message_t const * e)
{
    char const * result = 0;
    // looking for compiler name
    if ((e->cmd) && (e->cmd[0]) && match(&filter->compilers, e->cmd[0]))
    {
        // looking for source file
        char const * const * it = e->cmd;
        for (; *it; ++it)
        {
            if ((0 == result) && match(&filter->source_files, *it))
            {
                result = fix_path(*it, e->cwd);
            }
            else if (match(&filter->cancel_parameters, *it))
            {
                if (result)
                {
                    free((void *)result);
                    result = 0;
                }
                break;
            }
        }
    }
    return result;
}


static void compile(config_setting_t const * array, regex_list_t * prepared)
{
    prepared->total_count = 0;
    prepared->match_count = 0;
    prepared->length = config_setting_length(array);
    if (0 == prepared->length)
    {
        prepared->regexs = 0;
        return;
    }

    prepared->regexs = malloc(prepared->length * sizeof(regex_t));

    size_t idx = 0;
    regex_t * ot = prepared->regexs;
    for (; idx < prepared->length; ++idx, ++ot)
    {
        char const * const it = config_setting_get_string_elem(array, idx);
        int const result = regcomp(ot, it, REG_EXTENDED);
        if (0 != result)
        {
            size_t const errbuf_size = 256;
            char errbuf[errbuf_size];
            regerror(result, ot, errbuf, errbuf_size);
            fprintf(stderr, "bear: regcomp failed on '%s': %s\n", it, errbuf);
            exit(EXIT_FAILURE);
        }
    }
}

static int match(regex_list_t * prepared, char const * input)
{
    ++prepared->total_count;
    size_t idx = 0;
    for (; idx < prepared->length; ++idx)
    {
        regex_t * ot = prepared->regexs + idx;
        if (0 == regexec(ot, input, 0, 0, 0))
        {
            ++prepared->match_count;
            return 1;
        }
    }
    return 0;
}

static int is_empty(regex_list_t const * prepared)
{
    return (prepared->length == 0);
}

static void release(regex_list_t * prepared)
{
    size_t idx = 0;
    for (; idx < prepared->length; ++idx)
    {
        regex_t * ot = prepared->regexs + idx;
        regfree(ot);
    }
    free((void *)prepared->regexs);
}

static config_setting_t const * get_setting(config_setting_t const * config, char const * name)
{
    config_setting_t const * const result = config_setting_get_member(config, name);
    if (0 == result)
    {
        fprintf(stderr, "bear: could not find values for '%s' in file %s.\n",
                name,
                config_setting_source_file(config));
        exit(EXIT_FAILURE);
    }
    if (! config_setting_is_array(result))
    {
        fprintf(stderr, "bear: value for '%s' shall be array of strings in file %s at line %d.\n",
                name,
                config_setting_source_file(result),
                config_setting_source_line(result));
        exit(EXIT_FAILURE);
    }

    size_t const size = config_setting_length(result);
    size_t idx = 0;
    for (; idx < size; ++idx)
    {
        config_setting_t * const elem = config_setting_get_elem(result, idx);
        if (CONFIG_TYPE_STRING != config_setting_type(elem))
        {
            fprintf(stderr, "bear: value for '%s' shall be array of strings in file %s at line %d.\n",
                    name,
                    config_setting_source_file(elem),
                    config_setting_source_line(elem));
            exit(EXIT_FAILURE);
        }
    }
    return result;
}

static char const * fix_path(char const * file, char const * cwd)
{
    char * result = 0;
    if ('/' == file[0])
    {
        result = strdup(file);
        if (0 == result)
        {
            perror("bear: strdup");
            exit(EXIT_FAILURE);
        }
    }
    else
    {
        if (-1 == asprintf(&result, "%s/%s", cwd, file))
        {
            perror("bear: asprintf");
            exit(EXIT_FAILURE);
        }
    }
    return result;
}

File: rust/semantic/src/filter.rs
/*  Copyright (C) 2012-2024 by László Nagy
    This file is part of Bear.

    Bear is a tool to generate compilation database for clang tooling.

    Bear is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Bear is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::collections::HashSet;
use std::path::PathBuf;

use json_compilation_db::Entry;

/// A predicate that can be used to filter compilation database entries.
///
/// If the predicate returns `true`, the entry is included in the result set.
/// If the predicate returns `false`, the entry is excluded from the result set.
pub type EntryPredicate = Box<dyn FnMut(&Entry) -> bool>;

/// Represents a builder object that can be used to construct an entry predicate.
///
/// The builder can be used to combine multiple predicates using logical operators.
pub struct EntryPredicateBuilder {
    candidate: Option<EntryPredicate>,
}

impl std::ops::BitAnd for EntryPredicateBuilder {
    type Output = EntryPredicateBuilder;

    fn bitand(self, rhs: Self) -> Self::Output {
        match (self.candidate, rhs.candidate) {
            (None, None) =>
                EntryPredicateBuilder::new(),
            (Some(mut lhs), Some(mut rhs)) => {
                EntryPredicateBuilder::from(move |entry| {
                    let result = lhs(entry);
                    if result {
                        rhs(entry)
                    } else {
                        result
                    }
                })
            }
            (None, some) =>
                EntryPredicateBuilder { candidate: some },
            (some, None) =>
                EntryPredicateBuilder { candidate: some },
        }
    }
}

impl std::ops::Not for EntryPredicateBuilder {
    type Output = EntryPredicateBuilder;

    fn not(self) -> Self::Output {
        match self.candidate {
            Some(mut original) => {
                Self::from(move |entry| {
                    let result = original(entry);
                    !result
                })
            }
            None =>
                Self::new(),
        }
    }
}

impl EntryPredicateBuilder {
    /// Creates an entry predicate from the builder.
    pub fn build(self) -> EntryPredicate {
        match self.candidate {
            Some(predicate) => predicate,
            None => Box::new(|_: &Entry| true),
        }
    }

    /// Construct a predicate builder that is empty.
    #[inline]
    fn new() -> Self {
        Self { candidate: None }
    }

    /// Construct a predicate builder that implements a predicate.
    #[inline]
    fn from<P>(predicate: P) -> Self
    where
        P: FnMut(&Entry) -> bool + 'static,
    {
        Self { candidate: Some(Box::new(predicate)) }
    }

    /// Create a predicate that filters out entries
    /// that are using one of the given compilers.
    pub fn filter_by_compiler_paths(paths: Vec<PathBuf>) -> Self {
        if paths.is_empty() {
            Self::new()
        } else {
            Self::from(move |entry| {
                let compiler = PathBuf::from(entry.arguments[0].as_str());
                // return true if none of the paths are a prefix of the compiler path.
                paths.iter().any(|path| { !compiler.starts_with(path) })
            })
        }
    }

    /// Create a predicate that filters out entries
    /// that are using one of the given compiler arguments.
    pub fn filter_by_compiler_arguments(flags: Vec<String>) -> Self {
        if flags.is_empty() {
            Self::new()
        } else {
            Self::from(move |entry| {
                let mut arguments = entry.arguments.iter().skip(1);
                // return true if none of the flags are in the arguments.
                arguments.all(|argument| { !flags.contains(&argument) })
            })
        }
    }

    /// Create a predicate that filters out entries
    /// that are not using any of the given source paths.
    pub fn filter_by_source_paths(paths: Vec<PathBuf>) -> Self {
        if paths.is_empty() {
            Self::new()
        } else {
            Self::from(move |entry| {
                paths.iter().any(|path| { entry.file.starts_with(path) })
            })
        }
    }

    /// Create a predicate that filters out entries
    /// that source file does not exist.
    pub fn filter_by_source_existence(only_existing: bool) -> Self {
        if only_existing {
            Self::from(|entry| { entry.file.is_file() })
        } else {
            Self::new()
        }
    }

    /// Create a predicate that filters out entries
    /// that are already in the compilation database based on their hash.
    pub fn filter_duplicate_entries(hash_function: impl Fn(&Entry) -> u64 + 'static) -> Self {
        let mut have_seen = HashSet::new();

        Self::from(move |entry| {
            let hash = hash_function(entry);
            if !have_seen.contains(&hash) {
                have_seen.insert(hash);
                true
            } else {
                false
            }
        })
    }
}

#[cfg(test)]
mod test {
    use std::hash::{Hash, Hasher};
    use crate::{vec_of_pathbuf, vec_of_strings};
    use super::*;

    #[test]
    fn test_filter_by_compiler_paths() {
        let input: Vec<Entry> = vec![
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-c", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
            Entry {
                file: PathBuf::from("/home/user/project/source.c++"),
                arguments: vec_of_strings!["c++", "-c", "source.c++"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
            Entry {
                file: PathBuf::from("/home/user/project/test.c"),
                arguments: vec_of_strings!["cc", "-c", "test.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
        ];

        let expected: Vec<Entry> = vec![
            input[0].clone(),
            input[2].clone(),
        ];

        let sut: EntryPredicate = EntryPredicateBuilder::filter_by_compiler_paths(vec_of_pathbuf!["c++"]).build();
        let result: Vec<Entry> = input.into_iter().filter(sut).collect();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_filter_by_compiler_arguments() {
        let input: Vec<Entry> = vec![
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-c", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-cc1", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
            Entry {
                file: PathBuf::from("/home/user/project/test.c"),
                arguments: vec_of_strings!["cc", "-c", "test.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
        ];

        let expected: Vec<Entry> = vec![
            input[0].clone(),
            input[2].clone(),
        ];

        let sut: EntryPredicate = EntryPredicateBuilder::filter_by_compiler_arguments(vec_of_strings!["-cc1"]).build();
        let result: Vec<Entry> = input.into_iter().filter(sut).collect();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_filter_by_source_paths() {
        let paths_to_include = vec_of_pathbuf!["/home/user/project/source"];
        let paths_to_exclude = vec_of_pathbuf!["/home/user/project/test"];

        let input: Vec<Entry> = vec![
            Entry {
                file: PathBuf::from("/home/user/project/source/source.c"),
                arguments: vec_of_strings!["cc", "-c", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
            Entry {
                file: PathBuf::from("/home/user/project/test/source.c"),
                arguments: vec_of_strings!["cc", "-c", "test.c"],
                directory: PathBuf::from("/home/user/project"),
                output: None,
            },
        ];

        let expected: Vec<Entry> = vec![
            input[0].clone(),
        ];

        let sut: EntryPredicate = (
            EntryPredicateBuilder::filter_by_source_paths(paths_to_include) &
                !EntryPredicateBuilder::filter_by_source_paths(paths_to_exclude)
        ).build();
        let result: Vec<Entry> = input.into_iter().filter(sut).collect();
        assert_eq!(result, expected);
    }

    #[test]
    fn test_duplicate_detection_works() {
        let input: Vec<Entry> = vec![
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-c", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: Some(PathBuf::from("/home/user/project/source.o")),
            },
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-c", "-Wall", "source.c"],
                directory: PathBuf::from("/home/user/project"),
                output: Some(PathBuf::from("/home/user/project/source.o")),
            },
            Entry {
                file: PathBuf::from("/home/user/project/source.c"),
                arguments: vec_of_strings!["cc", "-c", "source.c", "-o", "test.o"],
                directory: PathBuf::from("/home/user/project"),
                output: Some(PathBuf::from("/home/user/project/test.o")),
            },
        ];

        let expected: Vec<Entry> = vec![
            input[0].clone(),
            input[2].clone(),
        ];

        let hash_function = |entry: &Entry| {
            let mut hasher = std::collections::hash_map::DefaultHasher::new();
            entry.file.hash(&mut hasher);
            entry.output.hash(&mut hasher);
            hasher.finish()
        };
        let sut: EntryPredicate = EntryPredicateBuilder::filter_duplicate_entries(hash_function).build();
        let result: Vec<Entry> = input.into_iter().filter(sut).collect();
        assert_eq!(result, expected);
    }
}